site/sfguides/src/production_ready_rag_with_trulens_snowflake_cortex/production_ready_rag_trulens_snowflake_cortex.md

```diff
@@ -0,0 +1,483 @@
+author: Josh Reini
+id: production_ready_rag_snowflake_trulens
+summary: This is a guide for building a production-ready RAG with Snowflake and TruLens.
+categories: Getting-Started
+environments: web
+status: Published
+feedback link: https://github.com/Snowflake-Labs/sfguides/issues
+tags: Getting Started, RAG, LLMs, TruLens, Snowflake
+
+# Production-Ready RAG with TruLens and Snowflake Cortex
+## Overview 
+Duration: 1
+
+In this quickstart, we'll show how to build a RAG with the full snowflake stack including [Cortex LLM Functions](https://docs.snowflake.com/en/user-guide/snowflake-cortex/llm-functions), [Cortex Search](https://github.com/Snowflake-Labs/cortex-search?tab=readme-ov-file), and [TruLens](https://www.trulens.org/) observability.
+
+In addition, we'll show how to run TruLens feedback functions with Cortex as the [feedback provider](https://www.trulens.org/trulens_eval/api/provider/), and how to log TruLens traces and evaluation metrics to a Snowflake table.
+
+Last, we'll show how to use [TruLens guardrails](https://www.trulens.org/trulens_eval/guardrails/) for filtering retrieved context and reducing hallucination.
+
+
+
+## Setup
+Duration: 2
+
+For this quickstart, you will need your Snowflake credentials and a Github PAT Token ready. For example purposes, we assume they are set in a `.env` file that looks like this:
+
+```
+# Loading data from github
+GITHUB_TOKEN=
+
+# Snowflake details
+SNOWFLAKE_USER=
+SNOWFLAKE_USER_PASSWORD=
+SNOWFLAKE_ACCOUNT=
+SNOWFLAKE_DATABASE=
+SNOWFLAKE_SCHEMA=
+SNOWFLAKE_WAREHOUSE=
+SNOWFLAKE_ROLE=
+SNOWFLAKE_CORTEX_SEARCH_SERVICE=
+```
+
+First, we'll install the packages needed:
+
+```python
+# pip install snowflake-snowpark-python
+# pip install notebook
+# pip install snowflake-ml-python
+# pip install trulens-eval
+# pip install snowflake-sqlalchemy
+# pip install llama-index
+# pip install llama-index-readers-github
+# pip install llama-index-embeddings-huggingface
+```
+
+Then we can load our credentials and set our Snowflake connection
+
+```python
+from dotenv import load_dotenv
+from snowflake.snowpark.session import Session
+import os
+
+load_dotenv()
+
+connection_details = {
+  'account':  os.environ["SNOWFLAKE_ACCOUNT"],
```

Comment:
there's an extra space here. Also you may want to keep the use of single vs double quotes consistent.